New package: qbox

EmreAtes opened the pull request.

adamjstewart commented on var/spack/repos/builtin/packages/qbox/package.py:
```diff
@@ -0,0 +1,95 @@
+        filter_file(r'\$\(TARGET\)', 'gcc_atlas', 'src/Makefile')
+
+    def install(self, spec, prefix):
+        os.mkdir(prefix + '/src')
```
You can replace this line with:
```python 
mkdir(prefix.src)
```

adamjstewart commented on var/spack/repos/builtin/packages/qbox/package.py:
```diff
@@ -0,0 +1,95 @@
+        filter_file(r'LIBS\s*\+=(.*-.*-.*)',  # Catch the line that's not fftw
+                    r'LIBS += -lsatlas \1 -lgfortran', mkfile)
+        shutil.move(mkfile, 'src/')
+        filter_file(r'\$\(TARGET\)', 'gcc_atlas', 'src/Makefile')
```
Some of this stuff looks like it will only work for GCC.

EmreAtes replied:
QBox requires atlas or BLAS, and spack couldn't install ATLAS with icc, pgi or clang that's available on my system.
I can change the dependency from atlas to openblas and try using that when installing qbox if it's necessary.

adamjstewart replied:
If that's the case, your package should depend on `blas`. Users can select whichever BLAS implementation they want on the command line:
```
$ spack install qbox ^intel-mkl
```
If you're having trouble installing ATLAS, you can try contacting the ATLAS developers and asking them to take a look. But ideally, the package should be made generic enough that it will work with any compiler and BLAS library.

adamjstewart commented on var/spack/repos/builtin/packages/qbox/package.py:
```diff
@@ -0,0 +1,95 @@
+
+    def install(self, spec, prefix):
+        os.mkdir(prefix + '/src')
+        shutil.move('src/qb', prefix + '/src/qb')
```
You can replace this line with:
```python 
install_tree('src/qb', prefix.src.qb)
```

adamjstewart commented on var/spack/repos/builtin/packages/comd/package.py:
```diff
@@ -50,7 +49,7 @@ def edit(self, spec, prefix):
 
     def install(self, spec, prefix):
-        shutil.move('bin', prefix)
+        install_tree('bin', prefix + '/bin')
```
You should use `prefix.bin` here.

adamjstewart commented on var/spack/repos/builtin/packages/qbox/package.py:
```diff
@@ -0,0 +1,108 @@
+
+    def edit(self, spec, prefix):
+        with open('src/spack.mk', 'w') as mkfile:
+            mkfile.write('CXX = {}\n'.format(spec['mpi'].mpicxx))
```
You'll need to use `{0}` instead of `{}` here. The latter doesn't work for Python 2.6

EmreAtes commented:
I removed the BLAS version dependency. I also modified an older package I had to use `install` instead of `shutil.move`.

adamjstewart commented on var/spack/repos/builtin/packages/qbox/package.py:
```diff
@@ -0,0 +1,108 @@
+            mkfile.write('CXXFLAGS = {}\n'.format(' '.join((
+                '-g', '-O3', '$(INCLUDE)', '$(DFLAGS)',
+            ))))
+        filter_file(r'\$\(TARGET\)', 'spack', 'src/Makefile')
```
If you don't want to bother with backslashes, you can add `string=True` to the end of this call. Spack will treat the first argument as a string instead of a regular expression.

adamjstewart commented on var/spack/repos/builtin/packages/qbox/package.py:
```diff
@@ -0,0 +1,108 @@
+                'XERCESC_3', 'MPICH_IGNORE_CXX_SEEK', 'SCALAPACK',
+                'USE_FFTW3', 'FFTWMEASURE', 'FFTW3_2D', 'ADD_',
+            ))))
+            mkfile.write('CXXFLAGS = {}\n'.format(' '.join((
```
`{0}`

davydden commented on var/spack/repos/builtin/packages/qbox/package.py:
```diff
@@ -0,0 +1,108 @@
+        with open('src/spack.mk', 'w') as mkfile:
+            mkfile.write('CXX = {0}\n'.format(spec['mpi'].mpicxx))
+            mkfile.write('LD = $(CXX)\n')
+            mkfile.write('LIBPATH = {0}\n'.format(' -L'.join((
```
That line won't work for mkl and others, use spec['blas'].libs (LibraryList) to get everything u need (see other packages with blas/scalapack)

davydden replied:
here's an excerpt from Abinit:
```
# BLAS/LAPACK/SCALAPACK-ELPA
        linalg = spec['lapack'].libs + spec['blas'].libs
        if '+scalapack' in spec:
            oapp('--with-linalg-flavor=custom+scalapack')
            linalg = spec['scalapack'].libs + linalg

        # elif '+elpa' in spec:
        else:
            oapp('--with-linalg-flavor=custom')

        oapp('--with-linalg-libs={0}'.format(linalg.ld_flags))
```

davydden commented on var/spack/repos/builtin/packages/qbox/package.py:
```diff
@@ -0,0 +1,108 @@
+            ))))
+            blas_names = spec['blas'].libs.names
+            mkfile.write('LIBS = {0}\n'.format(' -l'.join(
+                ['', 'fftw3',  'scalapack', 'xerces-c'] + blas_names
```
Don't hardcore 'scslapack'. U can get linking flags from LibraryList directly

davydden replied:
U can also add spec['blas'].libs+spec['scalapack'].libs etc and then ask for linking flags

adamjstewart commented:
> I think this is because there is no `libfftw.so` file in the `/lib` folder, because fftw libraries are called `libfftw3.so`. I don't know which part of what to change (perhaps the fftw package) to fix this, but I can include that fix in this PR as well, if you think it's necessary.

You would overwrite `libs` in the package. See `zlib` for an example.

EmreAtes commented:
I've modified `fftw` to locate `libfftw3.so` correctly. I don't know if any other ldflags are necessary, since the package installs a lot of `.so` objects, but only linking with `libfftw3.so` was enough for qbox. Also, I think `find_libraries` should also look for libraries with the `lib` prefix automatically, but I've added a `*` workaround that.

Also I've noticed that `$(INCLUDE)` wasn't actually necessary for compilation, even though it was there in the example Makefiles given by qbox.

davydden commented on var/spack/repos/builtin/packages/fftw/package.py:
```diff
@@ -70,6 +70,11 @@ class Fftw(AutotoolsPackage):
 
+    @property
+    def libs(self):
+        result = find_libraries(['*fftw3'], root=self.prefix, recurse=True)
```
What do u have as a result? Find_livbraries will be looking for 'lib*fftw3'

adamjstewart replied:
`find_libraries` accepts glob characters, but they really aren't needed here as `libfftw3` should always cut it.

davydden replied:
yes, that was my point as there's probably only `libfftw3` which satisfies `lib*fftw3`.

EmreAtes replied:
If I just write `fftw3`, it doesn't find `libfftw3.so` since `find_libraries` doesn't add the `lib` prefix ([see here](https://github.com/LLNL/spack/blob/9e8cce41e3f78ed4222eff342a5b61b6f8fbf645/lib/spack/llnl/util/filesystem.py#L971)). It just looks for `fftw.so` and fails to find it. I think using `libfftw3` instead would be better, since I checked the other uses of `find_libraries` and this is what they seem to do.
